src/driver-marsgaming/marsgaming-buttons.c

```diff
@@ -0,0 +1,425 @@
+};
+
+struct ratbag_button_action
+marsgaming_parse_button_to_action(struct ratbag_button *button,
+				  const struct marsgaming_button_info *button_info)
+{
+	const struct marsgaming_button_action_to_ratbag_parser *parser;
+	ARRAY_FOR_EACH(marsgaming_button_action_to_ratbag_parsers, parser) {
+		if (button_info->action == parser->marsgaming_action_id)
+			return parser->parse_action(button, button_info);
+	}
+	// If no action matches, set it to unknown
+	return (struct ratbag_button_action)BUTTON_ACTION_UNKNOWN;
+}
+
+enum marsgaming_button_id {
+	MARSGAMING_MM4_BUTTON_TOP_LEFT = 0,
+	MARSGAMING_MM4_BUTTON_TOP_RIGHT = 1,
+	MARSGAMING_MM4_BUTTON_TOP_MIDDLE = 2,
+	MARSGAMING_MM4_BUTTON_SIDE_1 = 3,
+	MARSGAMING_MM4_BUTTON_SIDE_2 = 4,
+	MARSGAMING_MM4_BUTTON_SIDE_3 = 5,
+	MARSGAMING_MM4_BUTTON_TOP_4 = 6,
+	MARSGAMING_MM4_BUTTON_TOP_6 = 7,
+	MARSGAMING_MM4_BUTTON_TOP_5 = 8,
+	MARSGAMING_MM4_BUTTON_BOTTOM_1 = 9,
+	MARSGAMING_MM4_BUTTON_SIDE_4 = 10,
+	MARSGAMING_MM4_BUTTON_SIDE_5 = 11,
+	MARSGAMING_MM4_BUTTON_SIDE_6 = 12,
+	MARSGAMING_MM4_BUTTON_SIDE_7 = 13,
+	MARSGAMING_MM4_BUTTON_SIDE_8 = 14,
+	MARSGAMING_MM4_BUTTON_SIDE_9 = 15,
+	MARSGAMING_MM4_BUTTON_SIDE_10 = 16,
+	MARSGAMING_MM4_BUTTON_SIDE_11 = 17,
+	MARSGAMING_MM4_BUTTON_SIDE_12 = 18,
+};
+
+struct marsgaming_button_mapping {
+	enum marsgaming_button_id button_id;
+	enum ratbag_button_type type;
+};
+
+static const struct marsgaming_button_mapping marsgaming_mm4_button_mapping[] = {
+	{ MARSGAMING_MM4_BUTTON_TOP_LEFT, RATBAG_BUTTON_TYPE_LEFT },
+	{ MARSGAMING_MM4_BUTTON_TOP_RIGHT, RATBAG_BUTTON_TYPE_RIGHT },
+	{ MARSGAMING_MM4_BUTTON_TOP_MIDDLE, RATBAG_BUTTON_TYPE_MIDDLE },
+	{ MARSGAMING_MM4_BUTTON_SIDE_1, RATBAG_BUTTON_TYPE_SIDE },
+	{ MARSGAMING_MM4_BUTTON_SIDE_2, RATBAG_BUTTON_TYPE_SIDE },
+	{ MARSGAMING_MM4_BUTTON_SIDE_3, RATBAG_BUTTON_TYPE_SIDE },
+	{ MARSGAMING_MM4_BUTTON_TOP_4, RATBAG_BUTTON_TYPE_EXTRA },
+	{ MARSGAMING_MM4_BUTTON_TOP_6, RATBAG_BUTTON_TYPE_RESOLUTION_DOWN },
+	{ MARSGAMING_MM4_BUTTON_TOP_5, RATBAG_BUTTON_TYPE_RESOLUTION_UP },
+	{ MARSGAMING_MM4_BUTTON_BOTTOM_1, RATBAG_BUTTON_TYPE_PROFILE_CYCLE_UP },
+	{ MARSGAMING_MM4_BUTTON_SIDE_4, RATBAG_BUTTON_TYPE_SIDE },
+	{ MARSGAMING_MM4_BUTTON_SIDE_5, RATBAG_BUTTON_TYPE_SIDE },
+	{ MARSGAMING_MM4_BUTTON_SIDE_6, RATBAG_BUTTON_TYPE_SIDE },
+	{ MARSGAMING_MM4_BUTTON_SIDE_7, RATBAG_BUTTON_TYPE_SIDE },
+	{ MARSGAMING_MM4_BUTTON_SIDE_8, RATBAG_BUTTON_TYPE_SIDE },
+	{ MARSGAMING_MM4_BUTTON_SIDE_9, RATBAG_BUTTON_TYPE_SIDE },
+	{ MARSGAMING_MM4_BUTTON_SIDE_10, RATBAG_BUTTON_TYPE_SIDE },
+	{ MARSGAMING_MM4_BUTTON_SIDE_11, RATBAG_BUTTON_TYPE_SIDE },
+	{ MARSGAMING_MM4_BUTTON_SIDE_12, RATBAG_BUTTON_TYPE_SIDE }
+};
+
+unsigned
+marsgaming_button_type(unsigned button_index)
+{
+	const struct marsgaming_button_mapping *mapping;
+	ARRAY_FOR_EACH(marsgaming_mm4_button_mapping, mapping) {
+		if (mapping->button_id == button_index)
+			return mapping->type;
+	}
+	return RATBAG_BUTTON_TYPE_UNKNOWN;
+}
```

Comment:
`enum ratbag_button_type` is deprecated so all of this is not useful, sorry.

Reply:
Should I not set ratbag_button::type then?

Reply:
Yes.
I believe it's not used anywhere anyway.